lib/geoip/geoip.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (C) 2024 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Package geoip provides an automatically updating MaxMind GeoIP2 database
+// provider.
+package geoip
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/oschwald/geoip2-golang"
+)
+
+const maxDatabaseSize = 1 << 30 // 1 GiB, at the time of writing the database is about 95 MiB
+
+type Provider struct {
+	edition         string
+	licenseKey      string
+	refreshInterval time.Duration
+	directory       string
+
+	mut        sync.Mutex
+	db         *geoip2.Reader
+	lastOpened time.Time
+}
+
+// NewGeoLite2CityProvider returns a new GeoIP2 database provider for the
+// GeoLite2-City database. The database will be stored in the given
+// directory (which should exist) and refreshed every 7 days.
+func NewGeoLite2CityProvider(licenseKey string, directory string) *Provider {
+	return &Provider{
+		edition:         "GeoLite2-City",
+		licenseKey:      licenseKey,
+		refreshInterval: 7 * 24 * time.Hour,
+		directory:       directory,
+	}
+}
+
+func (p *Provider) City(ip net.IP) (*geoip2.City, error) {
+	p.mut.Lock()
+
+	if p.db != nil && time.Since(p.lastOpened) > p.refreshInterval/2 {
+		p.db.Close()
+		p.db = nil
+	}
+	if p.db == nil {
+		var err error
+		p.db, err = p.open(context.Background())
+		if err != nil {
+			p.mut.Unlock()
+			return nil, err
+		}
+		p.lastOpened = time.Now()
+	}
+	db := p.db
+
+	p.mut.Unlock()
+
+	return db.City(ip)
+}
+
+// open returns a reader for the GeoIP2 database. If the database is not
+// available locally, it will be downloaded. If the database is older than
+// refreshInterval, it will be downloaded again. If the download fails, the
+// existing database will be used. The returned reader must be closed by the
+// caller in the normal manner.
+func (p *Provider) open(ctx context.Context) (*geoip2.Reader, error) {
+	if p.licenseKey == "" {
+		return nil, errors.New("open: no license key set")
+	}
```

Comment:
I don't remember, but I'm surprised this is now mandatory?

Reply:
It is...